lib/lower/ASTBuilder.cpp

```diff
@@ -0,0 +1,872 @@
+            [&](const Pa::EndAssociateStmt *) { e.isTarget = true; },
+            [&](const Pa::EndBlockStmt *) { e.isTarget = true; },
+            [&](const Pa::SelectCaseStmt *) {
+              e.setCFG(AST::CFGAnnotation::Switch, cstr);
+            },
+            [&](const Pa::CaseStmt *) { e.isTarget = true; },
+            [&](const Pa::EndSelectStmt *) { e.isTarget = true; },
+            [&](const Pa::EndChangeTeamStmt *) { e.isTarget = true; },
+            [&](const Pa::EndCriticalStmt *) { e.isTarget = true; },
+            [&](const Pa::NonLabelDoStmt *) {
+              e.isTarget = true;
+              e.setCFG(AST::CFGAnnotation::Iterative, cstr);
+            },
+            [&](const Pa::EndDoStmt *) {
+              e.isTarget = true;
+              e.setCFG(AST::CFGAnnotation::Goto, cstr);
+            },
+            [&](const Pa::IfThenStmt *) {
+              e.setCFG(AST::CFGAnnotation::CondGoto, cstr);
+            },
+            [&](const Pa::ElseIfStmt *) {
+              e.setCFG(AST::CFGAnnotation::CondGoto, cstr);
+            },
+            [&](const Pa::ElseStmt *) { e.isTarget = true; },
+            [&](const Pa::EndIfStmt *) { e.isTarget = true; },
+            [&](const Pa::SelectRankStmt *) {
+              e.setCFG(AST::CFGAnnotation::Switch, cstr);
+            },
+            [&](const Pa::SelectRankCaseStmt *) { e.isTarget = true; },
+            [&](const Pa::SelectTypeStmt *) {
+              e.setCFG(AST::CFGAnnotation::Switch, cstr);
+            },
+            [&](const Pa::TypeGuardStmt *) { e.isTarget = true; },
+            [&](const Pa::WhereConstruct *) {
+              // mark the WHERE as if it were a DO loop
+              e.isTarget = true;
+              e.setCFG(AST::CFGAnnotation::Iterative, cstr);
+            },
+            [&](const Pa::WhereConstructStmt *) {
+              e.setCFG(AST::CFGAnnotation::CondGoto, cstr);
+            },
+            [&](const Pa::MaskedElsewhereStmt *) {
+              e.isTarget = true;
+              e.setCFG(AST::CFGAnnotation::CondGoto, cstr);
+            },
+            [&](const Pa::ElsewhereStmt *) { e.isTarget = true; },
+            [&](const Pa::EndWhereStmt *) { e.isTarget = true; },
+            [&](const Pa::ForallConstructStmt *) {
+              e.isTarget = true;
+              e.setCFG(AST::CFGAnnotation::Iterative, cstr);
+            },
+            [&](const Pa::EndForallStmt *) { e.isTarget = true; },
+            [](const auto *) { /* do nothing */ },
+        },
+        e.u);
+  }
+}
+
+/// Annotate the AST with CFG source decorations (see CFGAnnotation) and mark
+/// potential branch targets
+inline void annotateFuncCFG(AST::FunctionLikeUnit &flu) {
+  annotateEvalListCFG(flu.evals, nullptr);
+}
+
+L::StringRef evalName(AST::Evaluation &e) {
+  return std::visit(
+      Co::visitors{
+          [](const Pa::AllocateStmt *) { return "AllocateStmt"; },
```

Comment:
`dump-parse-tree.h` also maps nodes to names. Maybe code could be shared?

Reply:
Thanks for pointing to this file, I have updated this code to use it (I had to make the `parser::ParseTreeDumper::GetNodeName` static on the way).